### PR TITLE
ATLAS-178: Support upload of marker images

### PR DESCRIPTION
**Pull request by HelioStrike**
JIRA issue: https://issues.openmrs.org/browse/ATLAS-178

The images are being stored on the MySQL server as a BLOB.

(I'm still looking for the CSS issue that's causing the image to float to the left, instead having its own row) 
Screenshot from 2019-07-12 18-10-35
Screenshot from 2019-07-12 18-09-44
Screenshot from 2019-07-12 18-20-23
Screenshot from 2019-07-12 18-20-31

### CODE REVIEW

**Comment by HelioStrike** [comment]
@bmamlin @cintiadr I just realized that the way I changed the database may be troublesome for the RSS feed if we'd like to display image links. I think we'll need an API to retrieve images to keep everything consistent. What do you think?

File: public/js/user.js [hunk 1]
```diff
@@ -340,7 +373,8 @@ function updateMarker(id) {
       contentType: "application/json",
     })
         .done(function(response) {
-          sites[id].siteData = response; 
+          sites[id].siteData = response;
+          sites[id].siteData.image = bufferToString(sites[id].siteData.image);
```
Comment (bmamlin) [inline public/js/user.js]: This didn't appear to be fully working for me (testing in Chrome). I had to manually reload the page before the marker's bubble showed the image.

**Comment by HelioStrike** [comment]
@bmamlin After an hour of searching, I finally figured out how to send data strings as images. -_-; While fixing up the remaining code to use the API, I feel like it would be easier to have a new table for images. It would make the code more readable, and we won't need to alter the image column manually everytime a GET is being performed. If the image is stored in the atlas table, we may also need to edit existing queries to avoid huge image reads . Additionally, if we create a new table, markers currently using external links won't be affected as the table and the way atlas(image) remains the same. What do you think of this?

**Comment by HelioStrike** [comment]
@bmamlin After an hour of searching, I finally figured out how to send data strings as images. -_-; While fixing up the remaining code to use the API, I feel like it would be easier to have a new table for images (say atlas_images(id, datastring)). It would make the code more readable, and we won't need to alter the image column manually everytime a GET is being performed. If the image is stored in the atlas table, we may also need to edit existing queries to avoid huge image reads . Additionally, if we create a new table, markers currently using external links won't be affected as the table and the way atlas(image) remains the same. If you feel like the image column is just wasted memory, I guess we can remove it from the atlas table, and still hold most of the previous properties. What do you think of this?

**Comment by HelioStrike** [comment]
@bmamlin I made the changes as requested :)

I guess we'll move on with the current plan for now. Maybe an another column can be added later to reduce read size each time, but if that sounds like a good idea, we could do it in another PR. Lets aim to get [ATLAS-184](https://github.com/openmrs/openmrs-contrib-atlas/pull/90) merged for now, once the other PRs get merged. :)

File: db/bootstrap.sql [hunk 2]
```diff
@@ -80,7 +80,7 @@ CREATE TABLE `atlas` (
   `name` varchar(1024) COLLATE utf8_unicode_ci NOT NULL,
   `url` varchar(1024) COLLATE utf8_unicode_ci DEFAULT NULL,
   `type` varchar(1024) COLLATE utf8_unicode_ci NOT NULL,
-  `image` varchar(1024) COLLATE utf8_unicode_ci DEFAULT NULL,
+  `image` BLOB NULL,
```
[the remark this answers is not in the document]
Comment (HelioStrike) [reply db/bootstrap.sql]: Thank you for this code!!! I haven't really written a lot of SQL of this kind, but now that I have this code, I'm pretty sure I can learn and extend it forward. :)

I sort of misinterpreted what `update.sql` was. I'll update the database cleanup PR once the other PRs get merged.

File: public/js/user.js [hunk 3]
```diff
@@ -406,10 +437,12 @@ function contentInfowindow(site) {
 
 function contentEditwindow(site) {
   var html = "<div class='site-bubble bubble-form'>";
-  html += "<form method='post' id='"+ site.id +"'>";
+  html += "<form method='post' id='"+ site.id +"' enctype='multipart/form-data'>";
   html += "<div class='form-group'><input type='text' required='true' placeholder='Site Name' title='Site Name' class='form-control input-sm' value='"+ site.name + "' id='name' name='name'></div>";
   html += "<div class='form-group'><input type='url' class='form-control input-sm' placeholder='Site URL' title='Site URL' value='"+ site.url + "' name='url' id='url'></div>";
-  html += "<div class='form-group'><input type='url' class='form-control input-sm' placeholder='Image' title='Image' value='"+ site.image + "' name='image' id='image'></div>";
+  html += "<div class='form-group'><p>Image: </p><input type='file' title='Image' name='image' id='image' onchange='readImage(event)'></div>";
+  html += "<div id='image-hint'>square (prefer 80x80) jpeg or png up to " + Math.round(maxImageUploadSize/1024) + " kb</div>"
```
Comment (bmamlin) [inline public/js/user.js]: Thanks for including this tip; however, I'd suggest tweaking the styling a little. Right now, it looks like this:

where the tip can be lost or feel like it belong with the contact name. I'd suggest styling it closer to something like this:

that is, put it between the "Image:" label and the file selector.

**Comment by HelioStrike** [comment]
@bmamlin I made the changes :)

```IF(image, concat(?,'://',?,'/marker/',id,'/image'), null) AS image_url,```

is an amazing idea, and completely eliminates the need for a new column! I had to tweak it to 

```IF(image is not null, concat(?,'://',?,'/marker/',id,'/image'), null) AS image_url,```

Its weird why the former doesn't work, but I thought I'd report the observation, as its something I may probably waste a lot of time on. :p

While testing, also take a look at the RSS feed, and the marker csv, as I made a few tiny changes to them while getting this PR workiing. :)

**Comment by HelioStrike** [comment]
How did this get closed? ._.

**Comment by HelioStrike** [comment]
Got closed accidentally, after I pushed without committing. Also, made the changes. :)

**Pull request merged**